File: backend/shared/lib/src/main/java/br/ufpr/dac/shared/dto/AutocadastroDto.java
package br.ufpr.dac.shared.dto;

import java.time.LocalDateTime;

import br.ufpr.dac.shared.dto.ContasDto.Conta;
import br.ufpr.dac.shared.dto.GerentesDto.Gerente;
import br.ufpr.dac.shared.dto.UsersDto.Cliente;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

public class AutocadastroDto {

  public enum StatusSolicitacao {
    PENDENTE,
    APROVADO,
    REJEITADO,
    FALHA
  }

  public enum TipoNotificacao {
    SOLICITACAO_RECEBIDA,
    APROVACAO,
    REJEICAO,
    FALHA
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class SolicitacaoEntrada {
    private Cliente cliente;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class Solicitacao {
    private Long id;
    private Cliente cliente;
    private Gerente gerente;
    private StatusSolicitacao status;
    private String motivoRejeicao;
    private LocalDateTime dataSolicitacao;
    private LocalDateTime dataAnalise;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class Aprovacao {
    private Long solicitacaoId;
    private String cpf;
    private Long gerenteId;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class Rejeicao {
    private Long solicitacaoId;
    private Long gerenteId;
    private String motivo;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class ContaCriada {
    private Long solicitacaoId;
    private Conta conta;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class UsuarioAuth {
    private Long solicitacaoId;
    private Long clienteId;
    private String email;
    private String senhaTemporaria;
    private String tipo;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class Notificacao {
    private Long solicitacaoId;
    private String destinatario;
    private String assunto;
    private String conteudoHtml;
    private TipoNotificacao tipo;
  }

  @Data
  @Builder
  @AllArgsConstructor
  @NoArgsConstructor
  public static class Falha {
    private Long solicitacaoId;
    private String cpf;
    private String email;
    private String motivo;
    private String operacaoOrigem;
  }

}

File: backend/apiGatewayServiceNode/src/dto/AutocadastroDto.ts
import { GerentesDtoGerente } from "./GerentesDto";
import { UsersDtoCliente } from "./UsersDto";

export type StatusSolicitacaoAutocadastro =
  | "PENDENTE"
  | "APROVADO"
  | "REJEITADO"
  | "FALHA";

export type TipoNotificacaoAutocadastro =
  | "SOLICITACAO_RECEBIDA"
  | "APROVACAO"
  | "REJEICAO"
  | "FALHA";

export interface AutocadastroSolicitacaoEntrada {
  cliente: UsersDtoCliente;
}

export interface AutocadastroSolicitacao {
  id: null | number;
  cliente: null | UsersDtoCliente;
  gerente: null | GerentesDtoGerente;
  status: null | StatusSolicitacaoAutocadastro;
  motivoRejeicao: null | string;
  dataSolicitacao: null | string;
  dataAnalise: null | string;
}

export interface AutocadastroAprovacao {
  solicitacaoId: null | number;
  cpf: null | string;
  gerenteId: null | number;
}

export interface AutocadastroRejeicao {
  solicitacaoId: number;
  gerenteId: null | number;
  motivo: string;
}

export interface AutocadastroContaCriada {
  solicitacaoId: number;
  conta: unknown;
}

export interface AutocadastroUsuarioAuth {
  solicitacaoId: number;
  clienteId: number;
  email: string;
  senhaTemporaria: string;
  tipo: "CLIENTE";
}

export interface AutocadastroNotificacao {
  solicitacaoId: null | number;
  destinatario: string;
  assunto: string;
  conteudoHtml: string;
  tipo: TipoNotificacaoAutocadastro;
}

export interface AutocadastroFalha {
  solicitacaoId: null | number;
  cpf: null | string;
  email: null | string;
  motivo: string;
  operacaoOrigem: string;
}
